### DISCUSSION TITLE
Autofixture.AutoMoq Registering Interface

### DISCUSSION THREAD

**Discussion by SaurabhHarwande** [AutoFixture/AutoFixture#1102] [Q&A]
I have just started exploring AutoFixture with xUnit2 and AutoMoq. I am using AutoMoqDataAttribute to inject the objects into the test methods.
I am using interfaces to register the Object to interface type. Following is the code for same.
```C#
public void GetAll_Success([Frozen]Mock<IValueRepository> mockValueRepository, IValueLogic objValueLogic) {
    mockValueRepository.Setup(x => x.GetAll()).Returns(new string[] {"Value1", "Value2"});
    IEnumerable<string> expected = new string[] {"Value1", "Value2"};
    IEnumerable<string> actual = objValueLogic.GetAll();
    Assert.Equal<IEnumerable<string>>(expected, actual);
}
```
```C#
public class AutoMoqDataAttribute :
    AutoDataAttribute {
    public AutoMoqDataAttribute() :
        base(() => {
            IFixture objFixture = new Fixture()
                .Customize(new AutoMoqCustomization());
            objFixture.Register<IValueLogic>(() => objFixture.Create<ValueLogic>());
            return objFixture;
        }) {}
}
```
Now this works alright as a new ValueLogic object is created and injected whenever we need it in test function.
The issue arises when I need to mock the IValueLogic here. I think actual object is being created by the fixture instead of Mock object here.
```C#
public void Get_Success([Frozen]Mock<IValueLogic> mockValueLogic, ValuesController objValuesController) {
    mockValueLogic.Setup(x => x.Get(1)).Returns("Value1");
    string expected = "Value1";
    string actual = objValuesController.Get(1).Value;
    Assert.Equal(expected, actual);
}
```
What can be done differently in this case. Is it alright to register the IValueLogic in the class AutoMoqDataAttribute. Or is there a way to register it elsewhere.

**Comment by zvirja** [comment]
Hello,

If you customize your fixture globally and "teach" it how to create `IValueLogic` instances via `Register()` method, than indeed this customization overrides the default behavior for each time whenever interface is requested. Even though you freeze `Mock<IValueLogic>`, object of this type is never requested - the logic which would relay interface request to mock request is located after the customizations block.

Easy suggestion I could make here (except of removing the global customization) is to re-register the dependency:

```c#
public void Get_Success([Frozen]Mock<IValueLogic> mockValueLogic, IFixture fixture) {
    fixture.Inject(mockValueLogic.Object);
    ValuesController objValuesController = fixture.Create<ValuesController>();
    mockValueLogic.Setup(x => x.Get(1)).Returns("Value1");
    ....
}
```

It should do a trick 😉 

Other suggestions would require more effort, so I would consider them only if you need to solve this problem on a regular basis.

**Comment by SaurabhHarwande** [comment]
First of all, thanks for the answer. I did not know that we can simply inject the IFixture into the test function. So learned something new.
The solution you posted should surely work I have not tried it though. I wanted something where I don't have to repeat same piece of code for every function. I came up with something else in the mean time. Here is what I have changed.
Instead of using Register I went ahead and added a TypeRelay to Customizations(Don't really know if this is a good idea). I am passing two types into the constructor itself. Also there is one parameterless contructor I am using.
```C#
public AutoMoqDataAttribute(Type iType, Type cType) :
    base(() => {
        IFixture objFixture = new Fixture()
            .Customize(new AutoMoqCustomization());
        objFixture.Customizations.Add(new TypeRelay(iType, cType));
        return objFixture;
    }) {
}
public AutoMoqDataAttribute() :
    base(() => {
        return new Fixture()
            .Customize(new AutoMoqCustomization());
    }) {
}
```
And then passed on the respective types while using the attributes. Here is how it looks
```C#
[Theory, AutoMoqData(typeof(IValueLogic), typeof(ValueLogic))]
public void GetAll_Success([Frozen]Mock<IValueRepository> mockValueRepository, IValueLogic objValueLogic) {
    ...
}
```
The other method uses the class type directly so no need to pass in the types.
```C#
[Theory, AutoMoqData]
public void GetAll_Success([Frozen]Mock<IValueLogic> mockValueLogic, ValuesController objValuesController) {
    ...
}
```
Now this works as expected but would like to hear your thoughts. I have just started unit testing and using fixtures since yesterday so have no idea about what I really am doing.
Also as you said Other suggestions would require more effort but I surely would like to go through those as I am trying to learn.

**Comment by isberg** [comment]
Hello,
  this seems like the issue I have battled with the last few days. How to get the fixture to create my SUT using a mock that I can do verification on when using autodata.

```
[AutoData]
public void DoThing_ShouldCall_DoOnThing([FrozenMoq] Mock<IThing> thingMock, Stuff sut)
{
    sut.DoThing();
    thingMock.Verify(x => x.Do());
}
```
If there was a magic FrozenMoqAttribute (as used in the code above), then all would be well.

With some inspiration from [Mark Seemann](https://blog.ploeh.dk/2010/03/27/Freezingmocks/) and learning about creating your own CustomizeAttribute I came up with:

```
public class FrozenMoq : CustomizeAttribute
{
    public override ICustomization GetCustomization(ParameterInfo parameter)
    {
        return new FrozenMockCustomization(parameter);
    }
}
```

and with some effort and confusion

```
public class FrozenMockCustomization : ICustomization
{
    private ParameterInfo parameter;

    public FrozenMockCustomization(ParameterInfo parameter)
    {
        this.parameter = parameter;
    }

    public void Customize(IFixture fixture)
    {
        var mockType = parameter.ParameterType;
        var objectType = mockType.GenericTypeArguments.Single();

        Mock mock = (Mock)Activator.CreateInstance(mockType);

        fixture.Inject(mockType, mock);
        fixture.Inject(objectType, mock.Object);
    }
}
```

and to finally manage to simply inject the mock and the mock.Object in an easy way

```
public static class FixtureExtensions
{
    public static void Inject(this IFixture fixture, Type type, object instance)
    {
        var method = typeof(FixtureRegistrar).GetMethod("Inject");
        var genericMethod = method.MakeGenericMethod(type);
        genericMethod.Invoke(null, new[] { fixture, instance });
    }
}
```

Hope this help someone.
Btw I am using NUnit3 and AutoFixture.NUnit3, both version 4.8.

**Comment by Kralizek** [comment]
Why do you take Mock<T> as parameter instead of directly T?

You can then access the mock to perform setups and verifications via Mock.Get(mock).

Example
https://github.com/Nybus-project/Nybus/blob/b610f2b5aea40abbc1740e1a7e25e09d7a8b7708/tests/Tests.Nybus/NybusHostTests.cs#L59-L70
